include/fiction/technology/sidb_cluster_hierarchy.hpp

```diff
@@ -0,0 +1,1148 @@
+using sidb_cluster_ptr = std::shared_ptr<sidb_cluster>;
+/**
+ * A receptor state pairs the potential recepting cluster with the identifier of the SiDB.
+ */
+struct sidb_cluster_receptor_state
+{
+    /**
+     * Receptor cluster.
+     */
+    const sidb_cluster_ptr& cluster;
+    /**
+     * SiDB index. It is contained in the receptor cluster.
+     */
+    const uint64_t sidb_ix;
+};
+/**
+ * Forward declaration. Required for compilation due to the mutually recursive structure in this file.
+ */
+static inline uint64_t get_cluster_size(const sidb_cluster_ptr& c) noexcept;
+/**
+ * A projector state pairs the potential projecting cluster with the associated multiset charge configuration.
+ */
+struct sidb_cluster_projector_state
+{
+    /**
+     * Projector cluster.
+     */
+    const sidb_cluster_ptr& cluster;
+    /**
+     * Multiset charge configuration. It is an element of the charge space of the projector cluster.
+     */
+    const uint64_t multiset_conf;
+    /**
+     * Getter for the number of a given charge state in the multiset configuration.
+     *
+     * @tparam cs Charge state to count the number of occurrences in the projector state of.
+     * @return The number of occurrences of the given charge state in the multiset charge configuration. For a neutral
+     * charge, the number of occurrences is inferred by considering the size of the cluster in the projector state.
+     */
+    template <sidb_charge_state cs>
+    [[nodiscard]] constexpr inline uint64_t get_count() const noexcept
+    {
+        switch (cs)
+        {
+            case sidb_charge_state::NEGATIVE: return multiset_conf >> 32ull;
+            case sidb_charge_state::POSITIVE: return multiset_conf & 0xFFFFFFFF;
+            default: return get_cluster_size(cluster) - (multiset_conf >> 32ull) - (multiset_conf & 0xFFFFFFFF);
+        }
+    }
+};
+/**
+ * Projector state pointers are unique.
+ */
+using sidb_cluster_projector_state_ptr = std::unique_ptr<sidb_cluster_projector_state>;
+/**
+ * The electrostatic potential bounds required for the *Ground State Space* algorithm. As the domain in
+ * which our potential bounds live are simply the real numbers, we may think of the lower bound and upper bound domains
+ * to be separate partial order structures on the real number line, inverse to each other. The *Ground State Space*
+ * algorithm requires the properties of a lower semi-lattice for these domains, ie. all finite meets must exist. This is
+ * implemented for the lower and upper bound respectively simply by taking a minimum or a maximum. One may think of
+ * meets as follows, which is very relevant to intention of their application: a meet, or greatest lower bound, is the
+ * *maximal information* common to a set (of potential bounds). This semantic operation is essential to the *Ground
+ * State Space* algorithm, which thus envelops without loss of accuracy.
+ */
+enum class bound_direction : uint8_t
+{
+    LOWER = 0,
+    UPPER
```

Comment:
Could you add docstrings to the enum values?

Reply:
done